### DISCUSSION TITLE
Playing with Record's optional key in a List

### DISCUSSION THREAD

**Discussion by akhansari** [nushell/nushell#14216] [Q&A] [answered]
I have a json containing a list of objects.
But usually a Record's key may be missing.

When I do `open my.json | where optional_field =~ value`, I get "cannot find column".
When I do `open my.json | where optional_field? =~ value`, I get "nothing and string type mismatch".

So I'm forced to be very verbose like this:
```nushell
open my.json | where ($it.optional_field? | default "" | str contains value)
```

Is there any simpler solution?

**Comment by NotTheDr01ds** [comment, marked as answer]
Nushell's type checking does require a little bit more verboseness here, but there are some alternative forms you can try.

In all versions, I think `=~` (or the new `like` operator) are probably better options than `str contains`, but that's just preference.

```nu
# Sample data for demo purposes
let sample_json = [
  {
    "id": 1,
    "optional_field": "123value"
  },
  {
    "id": 2,
    "optional_field": "456"
  },
  {
    "id": 3
  }
]
```

Options:

### Option 1: `try` block to discard error results

The most concise, and beats out your original version by a few characters.  Just wraps the `str contains` (or `=~`) in a `try` to ignore errors:

```nu
$sample_json | where {try {$in.optional_field =~ value}}
# or
$sample_json | where {try {$in.optional_field | str contains value}}
# or
$sample_json | where {try {$in.optional_field like value}}
```

### Option 2: Double-filter

Same effect as the `and` version below.  This is my preference, as I think it's the most "readable" of the four options:

```
$sample_json
| where optional_field? != null
| where optional_field =~ value
```

As above, you can substitute `str contains` or `like` based on your preference.

### Option 3: An `and` condition

Perhaps more verbose, but possibly a bit more clear than the original version, at least.  I'd definitely use (1) or (2) above personally, but I know some people prefer the single `where`/`and` compared to a double-`where`.

```nu
$sample_json | where optional_field? != null and optional_field =~ value
# Alternatively, multi-line:
$sample_json | where (
  $it.optional_field? != null
) and (
  $it.optional_field =~ value
)
```

`and` short-circuits the operation for `null` fields and never runs the `=~` check.

**Reply by akhansari** [reply]
Thanks a lot for the explanation.
The option 3 seems elegant so I created a helper function for:

```nushell
def like [value: string] {
    $in != null and ($in | str contains -i $value)
}

open my.json | where ($it.optional_field? | like value)
```

**Reply by NotTheDr01ds** [reply]
Good idea - Just be aware that `like` will be an operator in the next release.  Probably not a great idea to name it that, but Nushell does allow it.  It's just that it can cause some issues if you ever run someone else's code in the future that uses the `like` operator.

**Reply by akhansari** [reply]
This is a great news! Especially for those who handle data frequently.
I'll rename my function name.

**Comment by ghost** [comment]
another not yet mentioned option is to just fill up all the non-existent fields beforehand:

```nu
open my.json | default "" optional_field | where optional_field =~ value
```
